plasmapy/atomic/nuclear.py

```diff
@@ -64,119 +65,235 @@ def nuclear_binding_energy(argument, mass_numb=None):
+
+        for particle in particles:
+            try:
+                baryon_number += mass_number(particle)
+            except ValueError:
+                if _is_antiproton(particle) or particle == 'antineutron':
+                    baryon_number -= 1
+
+        return baryon_number
+
+    def _total_charge(particles):
+        r"""Finds the total integer charge in a list of nuclides
+        (excluding bound electrons) and other particles."""
+
+        total_charge = 0
+
+        for particle in particles:
+            try:
+                total_charge += atomic_number(particle)
+            except ValueError:
+                total_charge += charge_state(particle)
+
+        return total_charge
+
+    def _mass_energy(species_list):
+        r"""Finds the total mass energy from a list of species, while
+        taking the masses of the fully ionized isotopes."""
+
+        total_mass = 0.0*units.kg
+
+        for species in species_list:
+            if _is_electron(species) or _is_positron(species):
+                total_mass += constants.m_e
+            elif _is_neutron(species):
+                total_mass += constants.m_n
+            elif _is_antiproton(species):
+                total_mass += constants.m_p
+            elif species == 'antineutron':
+                total_mass += constants.m_n
+            else:
+                atomic_numb = atomic_number(species)
+                total_mass += ion_mass(species, Z=atomic_numb)
+
+        return (total_mass * constants.c**2).to(units.J)
+
+    input_err_msg = ("The inputs to nuclear_reaction_energy should be either "
+                     "a string representing a nuclear reaction (e.g., "
+                     "'D + T -> He-4 + n') or the keywords 'reactants' and "
+                     "'products' as lists with the nucleons or particles "
+                     "involved in the reaction (e.g., reactants=['D', 'T'] "
+                     "and products=['He-4', 'n'].")
+
+    reaction_string_is_input = args and not kwargs and len(args) == 1
+
+    reactants_products_are_inputs = kwargs and not args and len(kwargs) == 2
+
+    if reaction_string_is_input == reactants_products_are_inputs:
+        raise ValueError(input_err_msg)
+
+    if reaction_string_is_input:
+
+        reaction = args[0]
+
+        if not isinstance(reaction, str):
+            raise TypeError(input_err_msg)
+        elif '->' not in reaction:
+            raise ValueError(f"The reaction '{reaction}' is missing a '->'"
+                             " or '-->' between the reactants and products.")
+
+        try:
+            LHS_string, RHS_string = re.split('-+>', reaction)
+            LHS_list = re.split(' \+ ', LHS_string)
+            RHS_list = re.split(' \+ ', RHS_string)
+            reactants = _get_species(LHS_list)
+            products = _get_species(RHS_list)
+        except Exception as ex:
+            raise ValueError(f"{reaction} is not a valid nuclear reaction.") \
+                from ex
```

Comment:
I haven't seen this `from ex` syntax before. What's that?

Reply:
I've never seen `from` being used for anything other than imports. This looks weird.

Reply:
I just learned this!  If you run this code snippet...
```Python
try:
    raise TypeError
except Exception as e:
    raise ValueError from e
```
then you get the following
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-42a31f7f16fe> in <module>()
      1 try:
----> 2     raise TypeError
      3 except Exception as e:

TypeError: 

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<ipython-input-5-42a31f7f16fe> in <module>()
      2     raise TypeError
      3 except Exception as e:
----> 4     raise ValueError from e
      5 

ValueError: 
```
Without the `as ex` and `from ex`, the 

```The above exception was the direct cause of the following exception:```

would instead be

```During handling of the above exception, another exception occurred:```

This is a way to help with tracebacks by keeping track of the root causes of exceptions.

Reply:
Well, that's pretty darn cool, then!